File: frontend/src/routes/BloodTests/lib/chartData.ts
import type { BloodTestRow } from '../schemas';

export type ChartDatum = {
  x: Date;
  y: number;
  timing: string;
  inRange: boolean | null;
  unit: string;
  datetime: string;
  refLow: number | null;
  refHigh: number | null;
};

export type ChartSeries = {
  id: string;
  data: ChartDatum[];
};

export type RefRange = {
  low: number;
  high: number;
  unit: string;
};

export function getReferenceRange(rows: BloodTestRow[]): RefRange | null {
  const withRange = rows
    .filter(r => r.ref_low != null && r.ref_high != null)
    .sort((a, b) => b.datetime.localeCompare(a.datetime));
  const latest = withRange[0];
  if (!latest || latest.ref_low == null || latest.ref_high == null) return null;
  return { low: latest.ref_low, high: latest.ref_high, unit: latest.unit };
}

export function getPointColor(datum: Pick<ChartDatum, 'timing'>): string {
  if (datum.timing === 'pre')  return '#22d3ee';  // pre-dialysis — cyan
  if (datum.timing === 'post') return '#f59e0b';  // post-dialysis — amber
  return '#818cf8';                               // plain / unknown timing — indigo
}

// Priority for same-day deduplication: pre beats post beats plain.
// This prevents pre+post readings on the same day creating a zigzag
// that looks like two separate lines.
const TIMING_RANK: Record<string, number> = { pre: 0, post: 1, '': 2 };

export function toNivoSeries(marker: string, rows: BloodTestRow[]): ChartSeries {
  // One representative reading per calendar date.
  const byDate = new Map<string, BloodTestRow>();
  for (const r of rows.filter(r => !r.qualitative)) {
    const dateKey = r.datetime.slice(0, 10);
    const existing = byDate.get(dateKey);
    const rank = TIMING_RANK[r.timing] ?? 2;
    if (!existing || rank < (TIMING_RANK[existing.timing] ?? 2)) {
      byDate.set(dateKey, r);
    }
  }

  const data: ChartDatum[] = Array.from(byDate.values())
    .sort((a, b) => a.datetime.localeCompare(b.datetime))
    .map(r => {
      const inRange =
        r.ref_low != null && r.ref_high != null
          ? r.value >= r.ref_low && r.value <= r.ref_high
          : null;
      return {
        x: new Date(r.datetime),
        y: r.value,
        timing: r.timing,
        inRange,
        unit: r.unit,
        datetime: r.datetime,
        refLow: r.ref_low,
        refHigh: r.ref_high,
      };
    });
  return { id: marker, data };
}

File: flutter/lib/features/blood_tests/chart_data.dart
import 'dart:ui';

import 'models.dart';

/// Port of lib/chartData.ts.

// pre = cyan, post = amber, plain = indigo (mode-agnostic accent hues).
const _preColor = Color(0xFF22D3EE);
const _postColor = Color(0xFFF59E0B);
const _plainColor = Color(0xFF818CF8);

Color pointColor(String timing) => switch (timing) {
      'pre' => _preColor,
      'post' => _postColor,
      _ => _plainColor,
    };

const lineColor = _plainColor;

/// Latest row that has a reference range (drives the shaded band).
RefRange? getReferenceRange(List<BloodTestRow> rows) {
  final withRange = rows
      .where((r) => r.refLow != null && r.refHigh != null)
      .toList()
    ..sort((a, b) => b.datetime.compareTo(a.datetime));
  if (withRange.isEmpty) return null;
  final r = withRange.first;
  return RefRange(r.refLow!, r.refHigh!, r.unit);
}

// Same-day dedup priority: pre beats post beats plain.
const _timingRank = <String, int>{'pre': 0, 'post': 1, '': 2};

/// One representative datum per calendar date, sorted ascending by datetime.
List<ChartDatum> toSeries(List<BloodTestRow> rows) {
  final byDate = <String, BloodTestRow>{};
  for (final r in rows.where((r) => !r.qualitative)) {
    final dateKey = r.datetime.substring(0, 10);
    final existing = byDate[dateKey];
    final rank = _timingRank[r.timing] ?? 2;
    if (existing == null || rank < (_timingRank[existing.timing] ?? 2)) {
      byDate[dateKey] = r;
    }
  }

  final sorted = byDate.values.toList()
    ..sort((a, b) => a.datetime.compareTo(b.datetime));

  return sorted.map((r) {
    final inRange = (r.refLow != null && r.refHigh != null)
        ? (r.value >= r.refLow! && r.value <= r.refHigh!)
        : null;
    return ChartDatum(
      dateMs: (DateTime.tryParse(r.datetime) ?? DateTime(2000))
          .millisecondsSinceEpoch
          .toDouble(),
      value: r.value,
      timing: r.timing,
      inRange: inRange,
      unit: r.unit,
      datetime: r.datetime,
      refLow: r.refLow,
      refHigh: r.refHigh,
    );
  }).toList();
}
